better parser & error handling

Plixo2 opened the pull request:
I implemented a Tokenizer to convert the source text into a list of `TokenRecords`, then apply these with a updated parser. 

These parse Expressions get then converted to the runtime expression

This type of Tokenizer and Parser doesn't copy any strings, only points to where the token is stored, so it's pretty fast.
It's possible to switch to a custom Tokenizer and Parser to take full advantage of this in the future.

I used [nom_locate](https://crates.io/crates/nom_locate) for some error reporting with source code location. And  [codespan_reporting](https://crates.io/crates/codespan-reporting) for expected errors (user errors)

The tokenizer is still missing support for string interpolation, but that can be implemented later

Plixo2 committed:
add tokens for parser

Plixo2 committed:
add debug prints

Plixo2 committed:
implemented new parser

Plixo2 committed:
fixed name

Plixo2 committed:
updated the parser for readability and added comments

Plixo2 committed:
updated 'and', 'or' and 'not' tokens and added comments to the lexer

Plixo2 committed:
updated comments

Plixo2 commented on mussel/src/main.rs:
```diff
@@ -55,12 +58,19 @@ fn main() -> Result<()> {
+    let expressions = parser2::parser(&result).map_err(|e| e.create_report(file, program))?;
+
+    expressions.iter().for_each(|r| println!("{:?}", r));
+
```
just print the tokens and the expression AST for now

Plixo2 commented on mussel/src/lexer.rs:
```diff
@@ -0,0 +1,277 @@
+    )(input)
+}
+
+/// Tests a string starting and ending with double quotes.
```
missing string interpolation here

Plixo2 commented on mussel/src/lexer.rs:
```diff
@@ -0,0 +1,277 @@
+}
+
+/// Tests a number literal.
+/// Matches the following regex: [0-9]+(\.[0-9]+)?
```
should use nom `double` for floats

Plixo2 commented on mussel/src/lexer.rs:
```diff
@@ -0,0 +1,277 @@
+        recognize(pair(digit1, opt(pair(char('.'), digit1)))),
+        |num_str: Span| {
+            if num_str.contains('.') {
+                Ok::<Token, ErrorTree<&'_ str>>(Token::Float)
```
should use nom `double` for floats

Plixo2 commented on mussel/src/lexer.rs:
```diff
@@ -0,0 +1,277 @@
+/// Tests a string starting and ending with double quotes.
+fn string_literal(input: Span) -> IResult<Token> {
+    let (input, _) = delimited(char('"'), take_while(|c| c != '"'), char('"'))(input)?;
+    Ok((input, Token::String))
```
missing string interpolation here

gianndev commented:
I took a look at the code and it looks like you're doing a great job. I especially like the fact that you're implementing a lexer using a new library: this definitely makes Mussel much more powerful and similar to "big" programming languages. At the moment the PR is set as draft and I can't wait to see the final version!

Plixo2 commented:
Thank you. I will complete it today or tomorrow

Plixo2 committed:
Merge branch 'develop' of https://github.com/gianndev/mussel into dev_better_parser

Plixo2 committed:
updated to newer errors

Plixo2 committed:
Merge remote-tracking branch 'upstream/develop' into dev_better_parser

# Conflicts:
#	mussel/Cargo.toml
#	mussel/src/error.rs
#	mussel/src/main.rs

Plixo2 committed:
merge changes from develop branch

Plixo2 committed:
extracted the expressions into a new file

gianndev commented:
I'm noticing that you're making a lot of major changes to the Mussel code, and since so far compared to version _0.1.0_, a lot of changes have already been made, if you agree I would release version _0.2.0_ with the current code.

In this way we can also publicise Mussel by saying that there has recently been a new release. .

MatyasZsombor commented:
Does your current parser implementation support resynchronization? If not, I would be happy to implement it.

Plixo2 commented:
@gianndev no problem. Since my university has picked up again, i don't know when the parser will be ready. Would you consider creating new branch, after my next commit this evening, so others (@MatyasZsombor ) can work on it too?

The only things left are:

- converting from the expressions from the new parser to the Expressions needed for runtime. This should be relatively trivial. 

- improving float parsing. Replace the number tokenizer with the build-in nom `double`. 

- handle string interpolation: I never did this, but apparently you maintain a stack with the type of curly braces you encounter along with a state machine. Encountering a string would transition into the string state. When you encounter a opening brace, push a starting-interpolation-brace onto the stack and transition back to the normal mode. When you encounter a opening brace in normal mode, push a start-default token onto the stack. When encountering a closing brace, match the type of opening brace and when the top is a starting-interpolation-brace, transition back to the string state


@MatyasZsombor I didn't and I don't know how to do it properly in nom, so help would be appreciated

gianndev commented:
> Would you consider creating new branch, after my next commit this evening, so others (@MatyasZsombor) can work on it too?

Do you mean that tonight you're going to send me a PR and that I should create a branch after accepting the changes in the branch `develop`? Because in this case I think @MatyasZsombor can just update the branch `develop` of its fork. 

Then of course if @MatyasZsombor has any problems just tell me.

Plixo2 committed:
updated imports & completed conversion to runtime expressions

Plixo2 committed:
completed new parser implementation

Plixo2 committed:
updated error handling & number parsing

Plixo2 committed:
removed old parser

Plixo2 committed:
removed unused imports & renamed parser

Plixo2 committed:
fixed string end and starts

Plixo2 commented:
Everything is working now, please review it

gianndev reviewed (changes requested):
I tried to run some Mussel code but I get printed on the screen as output not only what the code requires but also other information that maybe are fine in the development phase, but are not suitable for production. Let me explain: I tried to run the _hello.mus_ code, which simply prints a "Hello, World!" on the screen, but instead of having only that text I got printed on the screen:

```
 Finished `dev` profile [unoptimized + debuginfo] target(s) in 0.15s
     Running `target\debug\mussel.exe ../examples/hello.mus`
TokenRecord { token_type: Identifier, offset: 0, length: 7 }
TokenRecord { token_type: LParenthesis, offset: 7, length: 1 }
TokenRecord { token_type: String, offset: 8, length: 16 }
TokenRecord { token_type: RParenthesis, offset: 24, length: 1 }
Call { region: TokenRecord { token_type: LParenthesis, offset: 7, length: 1 }, left: Identifier(TokenRecord { token_type: Identifier, offset: 0, length: 7 }), args: [String(TokenRecord { token_type: String, offset: 8, length: 16 })] }
Hello, Mussel!
```

**I think it would be appropriate to avoid showing all this information on the screen in the terminal when you are trying to run simple code.**

As for the new error handling I am very happy that now they are finally shown with a real, clear and direct error message.

Plixo2 commented:
Yes, but you can remove that yourself. I have to focus on my university now.

Just remove the prints in the main file.

gianndev commented:
Ok i perfectly understand you're busy now. I'll correct it by myself, and **thanks again for have made Mussel a really better language** 😁

gianndev committed:
Removed unuseful prints

I deleted some parts of the code to avoid printing too many unuseful information when a Mussel code is running

gianndev reviewed (approved):
Everything is fine, thanks again for the effort in upgrading Mussel so much
